Add settings property to optimizer classes in ``qiskit.algorithms.optimizers``

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
 
Enable dumping `qiskit.algorithms.optimizers.Optimizer` objects to JSON-serializable dictionaries and load from these.
The main motivation is that this will allow sending the optimizers to Qiskit Runtime programs.

Closes #6385.

mtreinish commented:
So a couple things, why does this need to exist in terra at all? If the ibm provider is going from an optimizer object in terra to a serialization format for an http api that submits a job ibm's runtime why do we need to dictate that format in terra, ideally it should be the provider layer which adapts from the terra object to whatever format is required that api. If the interface around accessing the necessary parameters or instantiating instances isn't standardized or stable that's a bug. This is a big anti-pattern we've been working on fixing for a long time and is a large part of why we've moved away from qobj everywhere (which was the same thing as this).

The second thing is this isn't actually making anything serializeable it's just a dict representation of these objects which makes it slightly simpler to be serialized using the stdlib json module but there is nothing inherently un-serializeable about the objects as they exist today (which is kind of my larger question).

Cryoris commented:
I'd be fine with moving this to the provider, as I'm not aware of another use case than using the optimizer with Qiskit Runtime 🙂 

I don't quite get this point
> If the interface around accessing the necessary parameters or instantiating instances isn't standardized or stable that's a bug.

Do you mean the optimizers would have to have a consisting way of storing the internal settings?

jyu00 commented:
@mtreinish I agree that normally the communication format between ibmq-provider and the api should be determined and done in the provider. However, inputs to runtime programs are entirely defined by program writers and unknown to the provider, so it has no way of defining a schema. Right now we are just using pickle, which is less than ideal, but I don't know of another way to serialize arbitrary python objects (if you do I'm all ears). 

Therefore I think it makes more sense to define the serialization methods inside the classes, but perhaps they can be done in a more standard way, such as `to_json()`/`from_json()` or `__getstate__` and `__setstate__`. So it's not so much for communicating with ibmq but for cleaner serialization in general.

nonhermitian commented:
I still do not understand why this is needed at all. Everything that is needed is in the runtime images, so why not pass a flag that sets the optimizer in the runtime code?

Cryoris commented:
@nonhermitian that's exactly what we wanted to do with these dictionaries. But we want to specify not only the type/name of the optimizer, but also the settings. Therefore we wanted to pass a dictionary containing the optimizer name and settings.

The advantage of using dictionaries and not JSON (or any other format) is that users could specify the optimizers also without going through the `Optimizer` classes. For instance
```python
inputs = {'ansatz': circuit,
          'optimizer': {'name': 'GradientDescent', 'maxiter': 100, 'learning_rate': 0.1}}
provider.runtime.run(..., inputs=inputs, ...)
```

I'll happily change to another format or move the serialization to the provider (if we find a way to do that), but not being able to select the optimizer is quite a severe restriction on the runtime and I was hoping to include this into the 0.18 release. Maybe the current format is good enough for that? 🙂

nonhermitian commented:
Ok yeah, I should probably actually look at the code before I speak. This makes sense.

woodsp-ibm commented on test/python/algorithms/optimizers/test_optimizers.py:
```diff
@@ -155,5 +171,215 @@ def callback(x):
+
+        # fidelity cannot be serialized, so it must be added back in
+        serialized["fidelity"] = fidelity
+        reconstructed = Optimizer.from_dict(serialized)
```
Since the from_dict checks and raises an exception if fidelity is not there, should we test first without fidelity to assure that code, before its then added where we test it works.

woodsp-ibm commented on test/python/algorithms/optimizers/test_optimizers.py:
```diff
@@ -155,5 +171,215 @@ def callback(x):
+        with self.assertRaises(QiskitError):
+            _ = optimizer.to_dict()
+
+    def test_scipy_name_missing(self):
```
How about invalid name test if my name is incorrect or does not match an installed optimizer?

woodsp-ibm commented on test/python/algorithms/optimizers/test_optimizers.py:
```diff
@@ -155,5 +171,215 @@ def callback(x):
+    @data(
+        ("BFGS", {"maxiter": 100, "eps": np.array([0.1])}),
+        ("CG", {"maxiter": 100, "gtol": 1e-8}),
+        ("COBYLA", {"maxiter": 100}),
```
Hopefully maxiter of 100 is not some magic number - would be nice to see some other values as a bit more variety to the test.

Cryoris replied:
It shouldn't be 😉 I'll change some

Cryoris commented on qiskit/algorithms/optimizers/adam_amsgrad.py:
```diff
@@ -114,6 +114,21 @@ def __init__(
                 writer = csv.DictWriter(csv_file, fieldnames=fieldnames)
                 writer.writeheader()
 
+    def to_dict(self) -> Dict[str, Any]:
```
I updated `from_dict` to handle all cases in the base class. I thought this might be tricky since I wanted to allow users to omit the "name" if `from_dict` is called on a subclass (e.g. SPSA), but turns out it's not 👍🏻 

Concerning `to_dict`, I'm not sure if there's enough common structure to implement that in the base class. I'd be adding special implementations to probably half the classes anyways, or what do you think?

Cryoris commented on qiskit/algorithms/optimizers/bobyqa.py:
```diff
@@ -62,6 +63,22 @@ def get_support_level(self):
+        """Construct the optimizer from a dictionary."""
+        name = dictionary.pop("name", None)
+        if name is not None:
+            if name.lower() != "bobyqa":
```
Fixed by moving from_dict to the base class entirely.

woodsp-ibm commented on qiskit/algorithms/optimizers/optimizer.py:
```diff
@@ -212,6 +216,81 @@ def optimize(
+        name = dictionary["name"]
+
+        optimizers = {
+            "adam": opt.ADAM,
```
I had mentioned this to @Cryoris  but will note here. Can we not use the classname more directly as the name - since its just part of to_dict/from_dict and does not really need to be human readable/friendly, it would avoid having to have some big lookup table here that needs updating if we add/remove an optimizer. The class name can be relative to optimizer folder so it would not be a fully qualified class name allowing anything to go in the dict.

Cryoris replied:
I'm resolving this as the de-serialization will take place in the RuntimeDecoder, so we won't have a `from_dict` method here anymore, and I'll remove the `name` from the settings too, since the RuntimeEncoder can store the class name

mtreinish reviewed (changes requested):
Per our offline discussion we should refactor this to expose a common `settings` property to the optimizer class which will expose the necessary settings for each optimizer class instead of a `to_dict`/`from_dict` pair to each class.

Cryoris commented:
I've changed the PR according to our discussion and implemented the encoding & decoding in the runtime provider. Now I'm waiting on Qiskit/qiskit-ibmq-provider#977 to update the de/encoding which in turns is blocked on #6650 on our side.

woodsp-ibm commented on qiskit/algorithms/optimizers/adam_amsgrad.py:
```diff
@@ -114,6 +114,13 @@ def __init__(
 
+    @property
+    def settings(self) -> Dict[str, Any]:
+        settings = {"name": "ADAM"}
```
Is there are reason we do not just use the classname rather than coding a string directly - in the few I have seen this looks identical

Cryoris replied:
see resolved comment above 🙂

jyu00 commented on qiskit/algorithms/optimizers/adam_amsgrad.py:
```diff
@@ -114,6 +114,13 @@ def __init__(
+    @property
+    def settings(self) -> Dict[str, Any]:
+        settings = {"name": "ADAM"}
+        settings.update(self._options)
```
Given `set_options()` allows setting arbitrary options, I think it'd be safer to hard code the known option keywords rather than returning `self._options`.

Cryoris replied:
done 👍🏻

jyu00 commented on qiskit/algorithms/optimizers/gradient_descent.py:
```diff
@@ -132,6 +132,26 @@ def __init__(
 
+    @property
+    def settings(self) -> Dict[str, Any]:
+        if self.callback is not None:
+            raise ValueError("Cannot serialize GradientDescent with callback.")
```
We can keep `settings` generic and still return `callback` in the dictionary. The JSON encoder will just fail when it encounters a callable (although the error message won't be as obvious).

Cryoris replied:
yes, done 👍🏻

jyu00 commented on qiskit/algorithms/optimizers/optimizer.py:
```diff
@@ -212,6 +214,11 @@ def optimize(
         pass
 
+    @property
+    def settings(self) -> Dict[str, Any]:
```
Similar to the other PR, can we make this an abstract method so new child class will implement it?

Cryoris replied:
👍🏻

jyu00 commented on qiskit/algorithms/optimizers/scipy_optimizer.py:
```diff
@@ -85,6 +85,16 @@ def get_support_level(self):
+    @property
+    def settings(self) -> Dict[str, Any]:
+        settings = {
+            "name": self._method,
```
I think this should be `"method"` ?

Cryoris replied:
Now that we don't carry the `name` along anymore yeah, this should be `method`

jyu00 commented on releasenotes/notes/serializable-optimizers-642402828b0c733b.yaml:
```diff
@@ -0,0 +1,29 @@
+---
+features:
+  - |
+    Enable dumping :class:`~qiskit.algorithms.optimizers.Optimizer` objects to JSON-serializable
+    dictionaries and load from these. 
```
In addition to name changes (e.g. `settings` to `to_dict()`), I think we'd need to add a disclaimer here. The returned dictionary can be serialized to JSON only if the encoder they use supports all the types in it. For example, if `settings` contains a numpy array, which is not supported by the default `JSONEncoder `, then they need to use a custom encoder that supports it.

Cryoris replied:
good point, I updated the reno!

jyu00 reviewed (commented):
A few comments, mostly based on the ones I got from my own PR xD. As mentioned, you don't need to add `name` to `settings` since the encoder can just get it from the object.

jyu00 commented on qiskit/algorithms/optimizers/optimizer.py:
```diff
@@ -212,6 +240,11 @@ def optimize(
             )
         pass
 
+    @property
+    def settings(self) -> Dict[str, Any]:
+        """The optimizer settings in a dictionary format."""
+        raise NotImplementedError("Dictionary conversion not implemented for this optimizer.")
```
I think this needs to be removed since there's another `settings` at line 162

Cryoris replied:
yep, I missed that one, it should already be updated 🙂

jyu00 commented on qiskit/algorithms/optimizers/gsls.py:
```diff
@@ -97,6 +97,10 @@ def get_support_level(self) -> Dict[str, int]:
 
+    @property
+    def settings(self) -> Dict[str, Any]:
+        return self._options.copy()
```
Can you also make this return a dictionary with fixed keys instead of using `self._options`?

Cryoris replied:
I changed it to allow only the keys in `GSLS._OPTIONS` 👍🏻

jyu00 reviewed (commented):
A couple more minor comments but otherwise looks good!

woodsp-ibm commented on qiskit/algorithms/optimizers/optimizer.py:
```diff
@@ -147,6 +149,24 @@ def setting(self):
         return ret
 
+    @property
+    @abstractmethod
```
I guess this is a choice here - make it abstract to force all concrete sub-classes to have an implementation. Or leave it concrete here and raise an error. Reason I bring this up is because of the global optimizers we have on nlopt. I doubt nl opt native dependent library is built here in CI so the nlopt optimizer tests I imagine are skipped. At present I do not think any will work because a `settings` is not present at least here https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/algorithms/optimizers/nlopts/nloptimizer.py (or more likely all the sub-classes to make things work) as would be required if this was abstract

Cryoris replied:
Actually that still made the tests fail. But I agree it's a choice and we could also make it non-abstract. I added the settings to the nlopt optimizers for now.

mtreinish commented on qiskit/algorithms/optimizers/optimizer.py:
```diff
@@ -147,6 +149,24 @@ def setting(self):
         return ret
 
+    @property
+    @abstractmethod
```
Setting this as a required `abstractmethod` feels like a breaking change to me. Unless you don't expect any downstream code to implement custom optimizer classes that are built on this.

woodsp-ibm replied:
In this past this was just on the base class and gathered the attributes so you see a summary of the optimizer - it was used by VQE to summarize it configuration. I would not have expected in the past this to be overridden, although you could, because it pulled out all the variables automatically and included them. Now the meaning of this has changed somewhat too now, in that this is a selected sub-set that can be serialized and a sub-class would be expected to override this so things like callable that might be set, which are not serializable, are not included. It still gives a summary in the old sense so its kinda compatible in that regard. Its a choice whether to make this abstract in the base class, or not simply raise a runtime error should it be called on some future optimizer that does not have this. I think the choice was to go with abstract - I had mention this above https://github.com/Qiskit/qiskit-terra/pull/6545#discussion_r661583859

Cryoris replied:
But it's true that if a user implemented an `Optimizer` on their own then this would be a breaking change... Maybe it's better to not have it as abstract after all.

woodsp-ibm replied:
Having it abstract and an IDE can assist - if its not, and then its not implemented by some sub-class, it will fail at runtime later should the optimizer be serialized. Which is better to have going forwards - abstract or not? If a user has their own optimizer its not a difficult change for them - but if they do then it is a breaking change.

mtreinish replied:
Right, my preference is to not make this a required method and remove the `abstractmethod` because of the backwards compatibility issue with any existing downstream optimizer subclasses. The normal pattern for late additions to abstract class like this is to just raise `NotImplementedError` or do a new version with the extra attribute. This doesn't feel to me like something breaking any custom optimizer class over. Especially because the only place this is currently being used is by the ibmq-provider for the qiskit runtime service, so a custom optimizer doesn't need to be serialized for that at all since on the other side the runtime won't be able to instantiate an object from it with out having the custom class in the first place.

But, that being said if people feel super strongly that it needs to be there, I don't think it's necessarily a blocker to have `abstractmethod` set as long it's called out in the release notes as a breaking API change in the `upgrade` section.

Cryoris replied:
I made the settings non-abstract, after all the optimizers don't necessarily have to have this implement this, only when they should be sent to the runtime.

mtreinish reviewed (commented):
Overall this LGTM, but the use of an `abstractmethod` for the new property method feels like a breaking change. We either should remove the decorator (it shouldn't change anything as far as this PR is concerned) or document this as a potentially breaking change in an upgrade note.

mtreinish reviewed (approved):
LGTM, thanks for making the update

woodsp-ibm commented on qiskit/algorithms/optimizers/optimizer.py:
```diff
@@ -165,7 +164,7 @@ def settings(self) -> Dict[str, Any]:
 
         """
-        raise NotImplementedError
+        raise NotImplementedError("The settings method is not implemented per default.")
```
I think it could have had a better error msg - saying something like this can be implement to allow an optimizer to be serialized for th runtime. Anyway I am not going to hold this up because of that.
